Control/move_to_pose/move_to_pose_robot_class.py

```diff
@@ -0,0 +1,242 @@
+
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+from random import random
+import copy
+
+# Simulation parameters
+TIME_DURATION = 1000
+TIME_STEP = 0.01
+AT_TARGET_ACCEPTANCE_THRESHOULD = 0.01
+SHOW_ANIMATION = True
+
+simulation_running = True
+all_robots_are_at_target = False
+
+
+class Pose:
+    """2D pose"""
+
+    def __init__(self, x, y, theta):
+        self.x = x
+        self.y = y
+        self.theta = theta
+
+
+class PathFinderController:
+    """Path finder controller"""
+
+    def __init__(self, Kp_rho, Kp_alpha, Kp_beta):
+        self.Kp_rho = Kp_rho
+        self.Kp_alpha = Kp_alpha
+        self.Kp_beta = Kp_beta
+
+
+class Robot:
+    """Robot class"""
+
+    def __init__(self, name, color, max_linear_speed, max_angular_speed,
+                 path_finder_controller):
+        """Inits Robot using its general specifications"""
+        self.name = name
+        self.color = color
+        self.MAX_LINEAR_SPEED = max_linear_speed
+        self.MAX_ANGULAR_SPEED = max_angular_speed
+        self.path_finder_controller = path_finder_controller
+        self.x_traj = []
+        self.y_traj = []
+        self.pose = Pose(0, 0, 0)
+        self.pose_start = Pose(0, 0, 0)
+        self.pose_target = Pose(0, 0, 0)
+        self.is_at_target = False
+
+    def set_start_target_poses(self, pose_start, pose_target):
+        self.pose_start = copy.copy(pose_start)
+        self.pose_target = pose_target
+        self.pose = pose_start
+
+    def move(self, dt):
+        self.x_traj.append(self.pose.x)
+        self.y_traj.append(self.pose.y)
+
+        x_diff = self.pose_target.x - self.pose.x
+        y_diff = self.pose_target.y - self.pose.y
+        rho = np.hypot(x_diff, y_diff)
+
+        if(rho < AT_TARGET_ACCEPTANCE_THRESHOULD):
+            self.is_at_target = True
+
+        # Restrict alpha and beta (angle differences) to the range
+        # [-pi, pi] to prevent unstable behavior e.g. difference going
+        # from 0 rad to 2*pi rad with slight turn
+
+        alpha = (np.arctan2(y_diff, x_diff)
+                 - self.pose.theta + np.pi) % (2 * np.pi) - np.pi
+        beta = (self.pose_target.theta - self.pose.theta - alpha
+                + np.pi) % (2 * np.pi) - np.pi
+
+        linear_velocity = self.path_finder_controller.Kp_rho * rho
+        angular_velocity = (self.path_finder_controller.Kp_alpha * alpha
+                            + self.path_finder_controller.Kp_beta * beta)
+
+        if alpha > np.pi / 2 or alpha < -np.pi / 2:
+            linear_velocity = -linear_velocity
+
+        if abs(linear_velocity) > self.MAX_LINEAR_SPEED:
+            linear_velocity = (np.sign(linear_velocity)
+                               * self.MAX_LINEAR_SPEED)
+
+        if abs(angular_velocity) > self.MAX_ANGULAR_SPEED:
+            angular_velocity = (np.sign(angular_velocity)
+                                * self.MAX_ANGULAR_SPEED)
```

Comment:
I think this function has duplicated codes in ‘ move_to_pose.py’. I think some common function should be created in ‘ move_to_pose.py’, and this file should use these.

Reply:
No. the move_to_pose_robot_class.py and move_to_pose.py have totally different structures. I rewrote move_to_pose.py and turned it to move_to_pose_robot_class.py to make its functionalities more modular. I can create separate class module so that everyone can simply import it in his/her main script. Please let me know if you would like me to do that?

Reply:
I think you can move the `PathFinderController` to the `move_to_pose.py` and it has control logic. It is better taht this script imports it and uses it.